src/shims/panic.rs

```diff
@@ -0,0 +1,223 @@
+/// __rust_maybe_catch_panic
+///
+/// If a panic occurs, we update this data with
+/// the information from the panic site
+pub struct UnwindData<'tcx> {
+    pub data: Pointer<Tag>,
+    pub data_ptr: MPlaceTy<'tcx, Tag>,
+    pub vtable_ptr: MPlaceTy<'tcx, Tag>,
+    pub dest: PlaceTy<'tcx, Tag>,
+    pub ret: mir::BasicBlock,
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+
+    #[inline]
+    fn handle_rust_start_panic(&mut self, args: &[OpTy<'tcx, Tag>]) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        let start_panic_did = helpers::resolve_did(*this.tcx, &["panic_unwind", "__rust_start_panic"])?;
+        let start_panic_instance = ty::Instance::mono(*this.tcx, start_panic_did);
+        let start_panic_mir = this.load_mir(start_panic_instance.def, None)?;
+
+        this.push_stack_frame(
+            start_panic_instance,
+            DUMMY_SP,
+            start_panic_mir,
+            None,
+            StackPopCleanup::Goto { ret: None, unwind: None   }
+        )?;
+
+        // Copy first argument into new call frame
+        let mut new_args = this.frame().body.args_iter();
+        let orig_arg = this.read_scalar(args[0])?;
+        let new_arg = this.local_place(new_args.next().unwrap())?;
+        this.write_scalar(orig_arg, new_arg)?;
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn handle_miri_start_panic(
+        &mut self,
+        args: &[OpTy<'tcx, Tag>],
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        trace!("miri_start_panic: {:?}", this.frame().span);
+
+        if this.tcx.tcx.sess.panic_strategy() == PanicStrategy::Abort  {
+            // FIXME: Actually print out the payload here
+            throw_unsup_format!("the evaluated program panicked");
+        }
+
+        // Get the raw pointer stored in arg[0]
+        let scalar = this.read_immediate(args[0])?;
+        this.machine.panic_tmp_ptr = Some(scalar);
+
+
+        // Push a dummy 'trampoline' frame.
+        // This function does nothing - it simply allows us to
+        // push a frame to the call stack, so that we can start unwinding
+        // by popping a frame.
+
+        let trampoline_did = helpers::resolve_did(*this.tcx, &["panic_unwind", "miri_panic_trampoline"])?;
+        let trampoline_instance = ty::Instance::mono(*this.tcx, trampoline_did);
+        let trampoline_mir = this.load_mir(trampoline_instance.def, None)?;
+
+        this.push_stack_frame(
+            trampoline_instance,
+            DUMMY_SP,
+            trampoline_mir,
+            None,
+            StackPopCleanup::Goto { ret: None, unwind: None }
+        )?;
+
+        // Our 'stack_pop' method will see this flag, and signal
+        // to the interpreter that we should start unwinding
+        this.frame_mut().extra.is_panic_start = true;
```

Comment:
So, given that we know that this intrinsic is only called from `panic` in src/libpanic_unwind/miri.rs... why do we need this extra stack frame? What's wrong with doing `this.frame_mut().extra.is_panic_start = true;` *without* first pushing a new stack frame?

Reply:
That will indicate that we want to start unwinding when the function *calling* `miri_start_panic` returns. That is, `miri_start_panic` would 'return' normally, the `panic` function in `src/libpanic_unwind/miri.rs` would continue executing, and then we would 'magically' start unwinding.

Doing it this way makes the intrinsic work sanely - that is, we don't try to continue executing code which occurs after the intrinsic call.

Reply:
> That is, miri_start_panic would 'return' normally, the panic function in src/libpanic_unwind/miri.rs would continue executing, and then we would 'magically' start unwinding.

Exactly. That seems less magical than the trampoline to me.

We could document the effect of the intrinsic as "marks the current stack frame such that when it gets popped, unwinding starts". The return value of `miri_start_panic` should be `()` then, not `!`. Seems like a reasonable and understandable concept to me?

(Note that `imp::panic` has [return value `u32`](https://github.com/rust-lang/rust/blob/c3d2cbbc6af03467d0722e87b319df91adaa8332/src/libpanic_unwind/seh.rs#L229) for other platforms, so it doesn't seem like this always ends with a `!`.)

Reply:
> Exactly. That seems less magical than the trampoline to me.

I feel the opposite way - this seems extremely magical. The trampoline is an implementation detail that `libunwind_panic` doesn't need to be aware of (other than exporting it, but we could move that into Miri if we really wanted to).

I'm especially uncomfortable with the intrinsic explicitly making reference to 'the current stack frame'. What happens if the caller of the intrinsic gets inlined?

Making the intrinsic start unwinding immediately solves all of these problems - it does something perfectly well-defined (starts unwinding immediately), and prevents us from worrying about where exactly in `libpanic_unwind` it gets called.

Reply:
With the current implementation, this intrinsic is conceptually just a regular function (we could make it a foreign function, and the behavior would be unchanged). I feel that this is much easier to understand than an intrinsic whose behavior depends on the structure of the call stack.

Reply:
> I'm especially uncomfortable with the intrinsic explicitly making reference to 'the current stack frame'. What happens if the caller of the intrinsic gets inlined?

Hm. As a specification machine, the Rust Abstract Machine doesn't need to worry about inlining, and that's how I thought about Miri as well. But OTOH it might be nice to run Miri on optimized MIR, so this is a good argument.

I am still uncomfortable with this action-at-a-distance trampoline though. Instead of calling and returning, can we do some thing more immediately? `miri_start_panic` gets called with a `ret` and an `unwind` edge, can we basically jump directly to the `unwind` edge instead of indirecting through a stack frame push-pop to achieve the same?

Reply:
> Instead of calling and returning, can we do some thing more immediately

That would essentially be copying some of the logic from `handle_stack_pop`. The advantage of the trampoline is that it allows us to keep all of the logic for stopping and starting unwinding in the same function.

Reply:
How much logic would need copying? If you forward `unwind: Option<mir::BasicBlock>` to `handle_miri_start_panic`, isn't `self.goto_block(unwind.unwrap())` all you need to do?

Reply:
One advantage of my proposed approach is that we can entirely remove `is_panic_start` from the frame extra data -- that seems like a big plus in terms of simplicity to me. Instead of registering an action now to be executed later, we can just execute it now.

By jumping to an unwind block, we effectively are in "unwind mode" -- remember that there is no explicit tracking of that state, it is implicitly encoded through the basic block that we are currently executing.

Reply:
Now that I think about it, the real problem is `StackPopInfo::StartUnwinding`. If we don't start unwinding when we pop a frame (the trampoline), we either need a different way of signaling to Rust that unwidning has started, or implement any checks we do when unwinding starts entirely within Miri.

I think it's simpler to have the starting and stopping of unwinding handles by `pop_stack_frame` and `handle_stack_pop`, rather than splitting it over two places.

Reply:
"Starting unwinding" is just jumping to an unwind basic block, no more. Nothing needs to be signaled, and I think the implementation of an intrinsic called `miri_start_panic` is a much better place for "starting unwinding" than `pop_stack_frame`.

I agree that `StackPopInfo::StartUnwinding` is a problem -- if we can remove that enum variant (better the entire enum, but let's see about that), I'll be more happy.

Reply:
I've removed `StackPopInfo::StartUnwinding`, but kept the enum - it could be a boolean now, but I think having explicit variant names is much clearer.

Reply:
> I think having explicit variant names is much clearer.

Absolutely. Once this all lands I might come back to it and try to remove `StopUnwinding` as well, let's see. ;)